src/object/srv_ec_aggregate.c

```diff
@@ -2212,6 +2212,50 @@ agg_obj_is_leader(struct ds_pool *pool, struct daos_oclass_attr *oca,
 	return rc;
 }
 
+static int
+agg_filter(daos_handle_t ih, vos_iter_desc_t *desc, void *cb_arg, unsigned int *acts)
+{
+	struct ec_agg_param	*agg_param = (struct ec_agg_param *)cb_arg;
+	struct ec_agg_entry	*agg_entry = &agg_param->ap_agg_entry;
+	struct ec_agg_pool_info	*info = &agg_param->ap_pool_info;
+	struct daos_oclass_attr  oca;
+	int			 rc = 0;
+
+	if (desc->id_type != VOS_ITER_OBJ)
+		return 0;
+
+	rc = dsc_obj_id2oc_attr(desc->id_oid.id_pub, &info->api_props, &oca);
+	if (rc) {
+		D_ERROR("Skip object("DF_OID") with unknown class(%u)\n",
+			DP_OID(desc->id_oid.id_pub),
+			daos_obj_id2class(desc->id_oid.id_pub));
+		*acts = VOS_ITER_CB_SKIP;
+		agg_param->ap_credits++;
+		goto done;
+	}
+
+	if (!daos_oclass_is_ec(&oca)) { /* Skip non-EC object */
+		D_DEBUG(DB_EPC, "Skip oid:"DF_UOID" non-ec obj\n",
+			DP_UOID(desc->id_oid));
+		agg_param->ap_credits++;
+		*acts = VOS_ITER_CB_SKIP;
+	}
+done:
+	if (agg_param->ap_credits > agg_param->ap_credits_max) {
+		agg_param->ap_credits = 0;
+		D_DEBUG(DB_EPC, "EC aggregation yield type %d. acts %u\n",
+			desc->id_type, *acts);
+		if (!(*acts & VOS_ITER_CB_SKIP))
+			agg_reset_pos(desc->id_type, agg_entry);
```

Comment:
This is getting more and more tricky and bug prone now...

It was my bad to save the last processed entry in the pre callback because the initial iterator has only one callback, that resulted in all these dirty hacks: clear the last processed entry in certain circumstances to avoid skipping partially processed entry.. (imaging iterator yield in object level without descending into subtrees, the whole object will be skipped mistakenly if we don't reset the last processed object before yield).

Now w've already have two callbacks, pre callback is called before descending into subtrees and post callback is called after subtree processing done, so I think it's time to fix these dirty hacks by saving the last processed entry in the post callback. (probably it should be done in a separate patch)

Anyway, I think we shouldn't touch the 'last processed entry' in pre check callback.

Reply:
Ok, I can remove this